Ben/float 16

benleetownsend opened the pull request.

benleetownsend committed:
WIP: first run that fp16

ADD: first run at fp16 for bert based models

ADD: tcn fixes

ADD: switch to generator processing of sequence

ADD: switch to generator processing

FIX: remove generator processing

WIP:

benleetownsend committed:
ADD: cast predictions back to float32

benleetownsend committed:
ADD: speed improvements from sequence encoder improvements

ADD: exclude oscar from fp16 stuff

benleetownsend committed:
ADD: fp16 test and config default

FIX: config resolution

FIX: allow fp16 only for bert models

ADD: parity tests at float 16

FIX: Reset unused changes

madisonmay commented on tests/test_base_models.py:
```diff
@@ -241,6 +241,23 @@ def test_save_load(self):
+        model.fit(train_sample.Text, train_sample.Target)
+        predictions = model.predict(valid_sample.Text)
+        model.config.float_16_predict = False
+        # important to do it this way for the models that do not support fp16 as
```
Don't quite follow this comment. Could you explain further?

benleetownsend replied:
So when we explicitly set model.config.* it doesn't go through the config resolution step like if you set the config items properly. Only BERT and RoBERTa support the float_16 predict but in these cases a warning is raised and this config argument is set back to False. If we start with False and then use `model.config.float_16_predict = True` on any of the base models where it is not supported this will fail because the config resolution doesn't catch it.

benleetownsend replied:
So the real question I guess is why do we test the models that don't support it yet, but I think it's important to get defined behaviour in these cases so that someone with a GPU that supports it can just set this config argument for every model and not worry about getting undefined behaviour.

Another option would be to explicitly raise an exception in these cases and then change the tests accordingly.

We should maybe think about moving the config resolution into the config object itself though so that any time a conflicting set of config arguments are set this is resolved no matter how or when they are set.

lhz1029 replied:
yeah, moving to the config object seems like the cleanest approach.

madisonmay commented on tests/test_base_models.py:
```diff
@@ -241,6 +241,23 @@ def test_save_load(self):
+        new_predictions = model.predict(valid_sample.Text)
+        errors = 0
+        for prediction, new_prediction in zip(predictions, new_predictions):
+            if prediction != new_prediction:
```
Exact equality works here?!

lhz1029 replied:
Yeah I'm quite surprised too

benleetownsend replied:
Over a large enough dataset i'm sure a classification or two will come out differently, and for sequence labelling I have seen results vary by a token either way.

madisonmay replied:
Ah I was thinking about probas.  But these are just ints / labels

lhz1029 commented on tests/test_parity.py:
```diff
@@ -90,3 +90,30 @@ def test_roberta_featurize(self):
+                    'data/test-bert-activations.npy'
+                )
+            ),
+            atol=1e-1
```
That tolerance seems quite high. Is that what we expect?

benleetownsend replied:
It's the same tolerance we have for the main BERT/Roberta Models, but yes it is pretty high.

lhz1029 reviewed (approved):
For the most part LGTM! Out of curiosity, how did you go through the codebase to make sure you got everything? Just a search for all tf.get_variable, tf.cast, tf.zeros, and tf.ones?

benleetownsend commented:
@lhz1029 I changed all the variables with the overridden variable getter because we don't always have direct access to the variables themselves if some layers or keras module is used and then just searched for tf.float32 to find any explicit casts and changed these to cast to the dtype that it needs to be compatible with. Then set the default config to use float_16 on every inference call and ran the whole test suite to see where any implicit float32s were coming from.

madisonmay commented on finetune/model.py:
```diff
@@ -132,11 +142,17 @@ def _model_fn(features, labels, mode, params):
         pred_op = None
 
-        with tf.variable_scope(tf.get_variable_scope()):
+        if estimator_mode == tf.estimator.ModeKeys.PREDICT and params.float_16_predict:
```
Only request would be factoring this out into a separate fn to keep _model_fn relatively clean

benleetownsend committed:
ADD: ability to detect fp16 capable gpus

benleetownsend committed:
ADD: change default config option

lhz1029 commented on finetune/base.py:
```diff
@@ -317,6 +318,12 @@ def get_estimator(self, force_build_lm=False, build_explain=False, cache=False):
+            fp16_predict = self.config.float_16_predict
+            if fp16_predict:
+                if not gpu_info(config.session_config)["fp16_inference"]:
+                    fp16_predict = False
```
maybe add a warning here to the user?

madisonmay replied:
yeah would just be a matter of changing this to LOGGER.warning

lhz1029 commented on finetune/util/gpu_info.py:
```diff
@@ -0,0 +1,36 @@
+                else:
+                    LOGGER.warning(
+                        (
+                            "A GPU with float 16 support is available but it is not GPU:0,"
```
What happens in the case of multiple gpus, some with float 16 support and others without?

lhz1029 replied:
If it's too difficult to manage the various cases, we could just not support float16 unless all gpus that are visible have float16 enabled. Is that possible/easy to do?

benleetownsend replied:
By default predict is put into the first gpu. So if this is not float 16, we don't use float 16, this warning is just to let people know that.

lhz1029 replied:
:+1:

lhz1029 commented on finetune/config.py:
```diff
@@ -239,7 +239,7 @@ def get_default_config():
         # General Settings
         low_memory_mode=False,
-        float_16_predict=False,
+        float_16_predict=True,
```
oh we've decided to make this the default?

madisonmay replied:
I think let's make it opt in for a bit and then change the default once we've built some trust that it won't ever do us harm?  Have no reason to believe this should ever be problematic though given that the gpu_info check is in place.

benleetownsend committed:
ADD: log for float_16 being automatically turned off

madisonmay commented on finetune/model.py:
```diff
@@ -123,6 +163,7 @@ def target_model_op(featurizer_state, Y, params, mode, **kwargs):
         return target_model_state
 
     def _model_fn(features, labels, mode, params):
+        var_getter, features = get_variable_getter(mode, features, fp16_predict)
```
Sweet, thanks!

benleetownsend committed:
ADD: change default for predict float 16 to false

benleetownsend commented:
All tests pass.
